db/virtual_tables.cc

```diff
@@ -192,8 +192,19 @@ class token_ring_table : public streaming_virtual_table {
                 continue;
             }
 
-            std::vector<dht::token_range_endpoints> ranges = co_await _ss.describe_ring(e.name);
-            co_await emit_ring(result, e.key, "ALL", std::move(ranges));
+            if (_db.find_keyspace(e.name).get_replication_strategy().uses_tablets()) {
+                co_await _db.get_tables_metadata().for_each_table_gently([&, this] (table_id, lw_shared_ptr<replica::table> table) -> future<> {
+                    if (table->schema()->ks_name() != e.name) {
+                        co_return;
+                    }
+                    const auto& table_name = table->schema()->cf_name();
+                    std::vector<dht::token_range_endpoints> ranges = co_await _ss.describe_ring_for_table(e.name, table_name);
+                    co_await emit_ring(result, e.key, table_name, std::move(ranges));
```

Comment:
Now that Seastar has mandatory support for coroutines, we can think of using yields: instead of calling a function to emit a row, yield a row into the consumer.